File: j2kt/jre/java/native/java/lang/Character.java
/*
 *  Licensed to the Apache Software Foundation (ASF) under one or more
 *  contributor license agreements.  See the NOTICE file distributed with
 *  this work for additional information regarding copyright ownership.
 *  The ASF licenses this file to You under the Apache License, Version 2.0
 *  (the "License"); you may not use this file except in compliance with
 *  the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

package java.lang;

import static javaemul.internal.KtNativeUtils.ktNative;

import com.google.j2kt.annotations.HiddenFromObjC;
import javaemul.internal.annotations.KtName;
import javaemul.internal.annotations.KtNative;
import org.jspecify.annotations.NullMarked;
import org.jspecify.annotations.Nullable;

// TODO(b/223774683): Java Character should implement Serializable. Kotlin Char doesn't.
@KtNative(name = "kotlin.Char", companionName = "java.lang.Character")
@NullMarked
public final class Character implements Comparable<Character> {
  public static final char MIN_VALUE = '\u0000';

  public static final char MAX_VALUE = '\uFFFF';

  public static final int MIN_RADIX = 2;

  public static final int MAX_RADIX = 36;

  public static final Class<Character> TYPE = ktNative();

  public static final byte UNASSIGNED = 0;

  public static final byte UPPERCASE_LETTER = 1;

  public static final byte LOWERCASE_LETTER = 2;

  public static final byte TITLECASE_LETTER = 3;

  public static final byte MODIFIER_LETTER = 4;

  public static final byte OTHER_LETTER = 5;

  public static final byte NON_SPACING_MARK = 6;

  public static final byte ENCLOSING_MARK = 7;

  public static final byte COMBINING_SPACING_MARK = 8;

  public static final byte DECIMAL_DIGIT_NUMBER = 9;

  public static final byte LETTER_NUMBER = 10;

  public static final byte OTHER_NUMBER = 11;

  public static final byte SPACE_SEPARATOR = 12;

  public static final byte LINE_SEPARATOR = 13;

  public static final byte PARAGRAPH_SEPARATOR = 14;

  public static final byte CONTROL = 15;

  public static final byte FORMAT = 16;

  public static final byte PRIVATE_USE = 18;

  public static final byte SURROGATE = 19;

  public static final byte DASH_PUNCTUATION = 20;

  public static final byte START_PUNCTUATION = 21;

  public static final byte END_PUNCTUATION = 22;

  public static final byte CONNECTOR_PUNCTUATION = 23;

  public static final byte OTHER_PUNCTUATION = 24;

  public static final byte MATH_SYMBOL = 25;

  public static final byte CURRENCY_SYMBOL = 26;

  public static final byte MODIFIER_SYMBOL = 27;

  public static final byte OTHER_SYMBOL = 28;

  public static final byte INITIAL_QUOTE_PUNCTUATION = 29;

  public static final byte FINAL_QUOTE_PUNCTUATION = 30;

  public static final byte DIRECTIONALITY_UNDEFINED = -1;

  public static final byte DIRECTIONALITY_LEFT_TO_RIGHT = 0;

  public static final byte DIRECTIONALITY_RIGHT_TO_LEFT = 1;

  public static final byte DIRECTIONALITY_RIGHT_TO_LEFT_ARABIC = 2;

  public static final byte DIRECTIONALITY_EUROPEAN_NUMBER = 3;

  public static final byte DIRECTIONALITY_EUROPEAN_NUMBER_SEPARATOR = 4;

  public static final byte DIRECTIONALITY_EUROPEAN_NUMBER_TERMINATOR = 5;

  public static final byte DIRECTIONALITY_ARABIC_NUMBER = 6;

  public static final byte DIRECTIONALITY_COMMON_NUMBER_SEPARATOR = 7;

  public static final byte DIRECTIONALITY_NONSPACING_MARK = 8;

  public static final byte DIRECTIONALITY_BOUNDARY_NEUTRAL = 9;

  public static final byte DIRECTIONALITY_PARAGRAPH_SEPARATOR = 10;

  public static final byte DIRECTIONALITY_SEGMENT_SEPARATOR = 11;

  public static final byte DIRECTIONALITY_WHITESPACE = 12;

  public static final byte DIRECTIONALITY_OTHER_NEUTRALS = 13;

  public static final byte DIRECTIONALITY_LEFT_TO_RIGHT_EMBEDDING = 14;

  public static final byte DIRECTIONALITY_LEFT_TO_RIGHT_OVERRIDE = 15;

  public static final byte DIRECTIONALITY_RIGHT_TO_LEFT_EMBEDDING = 16;

  public static final byte DIRECTIONALITY_RIGHT_TO_LEFT_OVERRIDE = 17;

  public static final byte DIRECTIONALITY_POP_DIRECTIONAL_FORMAT = 18;

  public static final char MIN_HIGH_SURROGATE = '\uD800';

  public static final char MAX_HIGH_SURROGATE = '\uDBFF';

  public static final char MIN_LOW_SURROGATE = '\uDC00';

  public static final char MAX_LOW_SURROGATE = '\uDFFF';

  public static final char MIN_SURROGATE = MIN_HIGH_SURROGATE;

  public static final char MAX_SURROGATE = MAX_LOW_SURROGATE;

  public static final int MIN_SUPPLEMENTARY_CODE_POINT = 0x010000;

  public static final int MIN_CODE_POINT = 0x000000;

  public static final int MAX_CODE_POINT = 0X10FFFF;

  @HiddenFromObjC public static int SIZE = 16;

  // J2KT removed: Subset, UnicodeBlock

  public Character(char value) {}

  @KtName("toChar")
  public native char charValue();

  @Override
  public native int compareTo(Character c);

  public static native int compare(char lhs, char rhs);

  public static native Character valueOf(char c);

  public static native boolean isValidCodePoint(int codePoint);

  public static native boolean isSupplementaryCodePoint(int codePoint);

  public static native boolean isHighSurrogate(char ch);

  public static native boolean isLowSurrogate(char ch);

  public static native boolean isSurrogate(char ch);

  public static native boolean isSurrogatePair(char high, char low);

  public static native int charCount(int codePoint);

  public static native int toCodePoint(char high, char low);

  public static native int codePointAt(CharSequence seq, int index);

  public static native int codePointAt(char[] seq, int index);

  public static native int codePointAt(char[] seq, int index, int limit);

  public static native int codePointBefore(CharSequence seq, int index);

  public static native int codePointBefore(char[] seq, int index);

  public static native int codePointBefore(char[] seq, int index, int start);

  public static native int toChars(int codePoint, char[] dst, int dstIndex);

  public static native char[] toChars(int codePoint);

  public static native int codePointCount(CharSequence seq, int offset, int count);

  public static native int codePointCount(char[] seq, int offset, int count);

  public static native int offsetByCodePoints(CharSequence seq, int index, int codePointOffset);

  public static native int offsetByCodePoints(
      char[] seq, int start, int count, int index, int codePointOffset);

  public static native int digit(char c, int radix);

  @HiddenFromObjC
  public static native int digit(int codePoint, int radix);

  @Override
  public native boolean equals(@Nullable Object object);

  public static native char forDigit(int digit, int radix);

  @HiddenFromObjC
  public static native String getName(int codePoint);

  @HiddenFromObjC
  public static native int getNumericValue(char c);

  @HiddenFromObjC
  public static native int getNumericValue(int codePoint);

  @HiddenFromObjC
  public static native int getType(char c);

  @HiddenFromObjC
  public static native int getType(int codePoint);

  @HiddenFromObjC
  public static native byte getDirectionality(char c);

  @HiddenFromObjC
  public static native byte getDirectionality(int codePoint);

  @HiddenFromObjC
  public static native boolean isMirrored(char c);

  @HiddenFromObjC
  public static native boolean isMirrored(int codePoint);

  @Override
  public native int hashCode();

  public static native char highSurrogate(int codePoint);

  public static native char lowSurrogate(int codePoint);

  @HiddenFromObjC
  public static native boolean isAlphabetic(int codePoint);

  @HiddenFromObjC
  public static native boolean isBmpCodePoint(int codePoint);

  @HiddenFromObjC
  public static native boolean isDefined(char c);

  @HiddenFromObjC
  public static native boolean isDefined(int codePoint);

  public static native boolean isDigit(char c);

  public static native boolean isDigit(int codePoint);

  @HiddenFromObjC
  public static native boolean isIdentifierIgnorable(char c);

  @HiddenFromObjC
  public static native boolean isIdeographic(int codePoint);

  @HiddenFromObjC
  public static native boolean isIdentifierIgnorable(int codePoint);

  public static native boolean isISOControl(char c);

  @HiddenFromObjC
  public static native boolean isISOControl(int c);

  @HiddenFromObjC
  public static native boolean isJavaIdentifierPart(char c);

  @HiddenFromObjC
  public static native boolean isJavaIdentifierPart(int codePoint);

  @HiddenFromObjC
  public static native boolean isJavaIdentifierStart(char c);

  @HiddenFromObjC
  public static native boolean isJavaIdentifierStart(int codePoint);

  @Deprecated
  @HiddenFromObjC
  public static native boolean isJavaLetter(char c);

  @Deprecated
  @HiddenFromObjC
  public static native boolean isJavaLetterOrDigit(char c);

  @HiddenFromObjC
  public static native boolean isLetter(char c);

  @HiddenFromObjC
  public static native boolean isLetter(int codePoint);

  @HiddenFromObjC
  public static native boolean isLetterOrDigit(char c);

  @HiddenFromObjC
  public static native boolean isLetterOrDigit(int codePoint);

  @HiddenFromObjC
  public static native boolean isLowerCase(char c);

  @HiddenFromObjC
  public static native boolean isLowerCase(int codePoint);

  @Deprecated
  public static native boolean isSpace(char c);

  public static native boolean isSpaceChar(char c);

  public static native boolean isSpaceChar(int codePoint);

  @HiddenFromObjC
  public static native boolean isTitleCase(char c);

  @HiddenFromObjC
  public static native boolean isTitleCase(int codePoint);

  @HiddenFromObjC
  public static native boolean isUnicodeIdentifierPart(char c);

  @HiddenFromObjC
  public static native boolean isUnicodeIdentifierPart(int codePoint);

  @HiddenFromObjC
  public static native boolean isUnicodeIdentifierStart(char c);

  @HiddenFromObjC
  public static native boolean isUnicodeIdentifierStart(int codePoint);

  public static native boolean isUpperCase(char c);

  @HiddenFromObjC
  public static native boolean isUpperCase(int codePoint);

  public static native boolean isWhitespace(char c);

  public static native boolean isWhitespace(int codePoint);

  @HiddenFromObjC
  public static native char reverseBytes(char c);

  public static native char toLowerCase(char c);

  public static native int toLowerCase(int codePoint);

  @Override
  public native String toString();

  public static native String toString(char value);

  public static native String toString(int value);

  @HiddenFromObjC
  public static native char toTitleCase(char c);

  @HiddenFromObjC
  public static native int toTitleCase(int codePoint);

  public static native char toUpperCase(char c);

  public static native int toUpperCase(int codePoint);

  public static native int hashCode(char c);
}

File: j2kt/jre/java/native/java/lang/Character.kt
/*
 * Copyright 2022 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package java.lang

import javaemul.lang.codePointAt
import kotlin.jvm.javaPrimitiveType

/** java.lang.Character static method emulations */
class Character {
  companion object {
    operator fun invoke(c: Char): Char = c

    const val MIN_VALUE: Char = Char.MIN_VALUE

    const val MAX_VALUE: Char = Char.MAX_VALUE

    const val MIN_RADIX: Int = 2

    const val MAX_RADIX: Int = 36

    val TYPE: Class<Char> = Char::class.javaPrimitiveType!!

    const val DIRECTIONALITY_ARABIC_NUMBER: kotlin.Byte = 6
    const val DIRECTIONALITY_BOUNDARY_NEUTRAL: kotlin.Byte = 9
    const val DIRECTIONALITY_COMMON_NUMBER_SEPARATOR: kotlin.Byte = 7
    const val DIRECTIONALITY_EUROPEAN_NUMBER: kotlin.Byte = 3
    const val DIRECTIONALITY_EUROPEAN_NUMBER_SEPARATOR: kotlin.Byte = 4
    const val DIRECTIONALITY_EUROPEAN_NUMBER_TERMINATOR: kotlin.Byte = 5
    const val DIRECTIONALITY_LEFT_TO_RIGHT: kotlin.Byte = 0
    const val DIRECTIONALITY_LEFT_TO_RIGHT_EMBEDDING: kotlin.Byte = 14
    const val DIRECTIONALITY_LEFT_TO_RIGHT_OVERRIDE: kotlin.Byte = 15
    const val DIRECTIONALITY_NONSPACING_MARK: kotlin.Byte = 8
    const val DIRECTIONALITY_OTHER_NEUTRALS: kotlin.Byte = 13
    const val DIRECTIONALITY_PARAGRAPH_SEPARATOR: kotlin.Byte = 10
    const val DIRECTIONALITY_POP_DIRECTIONAL_FORMAT: kotlin.Byte = 18
    const val DIRECTIONALITY_RIGHT_TO_LEFT: kotlin.Byte = 1
    const val DIRECTIONALITY_RIGHT_TO_LEFT_ARABIC: kotlin.Byte = 2
    const val DIRECTIONALITY_RIGHT_TO_LEFT_EMBEDDING: kotlin.Byte = 16
    const val DIRECTIONALITY_RIGHT_TO_LEFT_OVERRIDE: kotlin.Byte = 17
    const val DIRECTIONALITY_SEGMENT_SEPARATOR: kotlin.Byte = 11
    const val DIRECTIONALITY_WHITESPACE: kotlin.Byte = 12
    const val DIRECTIONALITY_UNDEFINED: kotlin.Byte = -1

    const val MIN_HIGH_SURROGATE: Char = Char.MIN_HIGH_SURROGATE

    const val MAX_HIGH_SURROGATE: Char = Char.MAX_HIGH_SURROGATE

    const val MIN_LOW_SURROGATE: Char = Char.MIN_LOW_SURROGATE

    const val MAX_LOW_SURROGATE: Char = Char.MAX_LOW_SURROGATE

    const val MIN_SURROGATE: Char = Char.MIN_SURROGATE

    const val MAX_SURROGATE: Char = Char.MAX_SURROGATE

    const val MIN_SUPPLEMENTARY_CODE_POINT: Int = Char.MIN_SUPPLEMENTARY_CODE_POINT

    const val MIN_CODE_POINT: Int = Char.MIN_CODE_POINT

    const val MAX_CODE_POINT: Int = Char.MAX_CODE_POINT

    const val SIZE: Int = Char.SIZE_BITS

    fun valueOf(c: Char): Char = c

    fun compare(c1: Char, c2: Char): Int = c1.compareTo(c2)

    // TODO(b/367245215): Implement this method properly when Kotlin native supports it.
    fun getDirectionality(c: Char): kotlin.Byte = DIRECTIONALITY_UNDEFINED

    fun isDigit(c: Char): kotlin.Boolean = c.isDigit()

    // As Kotlin native doesn't seem to have support for digit checks for code points, we
    // assume false for code points outside of the 16 bit char range.
    fun isDigit(cp: Int): kotlin.Boolean = cp >= 0 && cp <= 0xffff && (cp as Char).isDigit()

    fun isISOControl(c: Char): kotlin.Boolean = c.isISOControl()

    fun isLetter(c: Char): kotlin.Boolean = c.isLetter()

    fun isLetterOrDigit(c: Char): kotlin.Boolean = c.isLetterOrDigit()

    fun isUpperCase(c: Char): kotlin.Boolean = c.isUpperCase()

    fun isLowerCase(c: Char): kotlin.Boolean = c.isLowerCase()

    fun isSpace(c: Char): kotlin.Boolean = "\t\n\u000C\r ".contains(c)

    // Exclude non-breaking spaces
    fun isWhitespace(c: Char): kotlin.Boolean =
      c.isWhitespace() && !"\u00a0\u2007\u202F".contains(c)

    fun isWhitespace(i: Int): kotlin.Boolean =
      if (i < Char.MAX_VALUE.code) isWhitespace(i.toChar())
      else toChars(i).concatToString().isBlank()

    // Exclude control characters
    fun isSpaceChar(c: Char): kotlin.Boolean =
      c.isWhitespace() && !"\t\n\u000B\u000C\r\u001C\u001D\u001E\u001F".contains(c)

    fun isSpaceChar(i: Int): kotlin.Boolean =
      if (i <= Char.MAX_VALUE.code) isSpaceChar(i.toChar())
      else toChars(i).concatToString().isBlank()

    fun isTitleCase(c: Char): kotlin.Boolean = c.isTitleCase()

    fun forDigit(digit: Int, radix: Int): Char = digit.digitToChar(radix).lowercaseChar()

    fun hashCode(c: Char): Int = c.hashCode()

    fun digit(ch: Char, radix: Int): Int = ch.digitToIntOrNull(radix) ?: -1

    fun charCount(codePoint: Int): Int = if (codePoint >= MIN_SUPPLEMENTARY_CODE_POINT) 2 else 1

    fun codePointAt(charArray: CharArray, index: Int, limit: Int = charArray.size): Int {
      val hiSurrogate: Char = charArray[index]
      var loSurrogate: Char = '\u0000'
      return if (
        isHighSurrogate(hiSurrogate) &&
          index + 1 < limit &&
          isLowSurrogate(charArray[index + 1].also { loSurrogate = it })
      )
        toCodePoint(hiSurrogate, loSurrogate)
      else hiSurrogate.code
    }

    fun codePointAt(charSequence: CharSequence, index: Int): Int {
      val hiSurrogate: Char = charSequence[index]
      var loSurrogate: Char = '\u0000'
      return if (
        isHighSurrogate(hiSurrogate) &&
          index + 1 < charSequence.length &&
          isLowSurrogate(charSequence[index + 1].also { loSurrogate = it })
      )
        toCodePoint(hiSurrogate, loSurrogate)
      else hiSurrogate.code
    }

    fun codePointBefore(cs: CharSequence, index: Int): Int {
      val loSurrogate = cs[index - 1]
      return if (loSurrogate.isLowSurrogate() && index > 1 && cs[index - 2].isHighSurrogate())
        toCodePoint(cs[index - 2], loSurrogate)
      else loSurrogate.code
    }

    fun codePointBefore(ca: CharArray, index: Int, start: Int = 0): Int {
      val loSurrogate = ca[index - 1]
      return if (
        loSurrogate.isLowSurrogate() && index - 2 >= start && ca[index - 2].isHighSurrogate()
      )
        toCodePoint(ca[index - 2], loSurrogate)
      else loSurrogate.code
    }

    fun codePointCount(seq: CharSequence, beginIndex: Int, endIndex: Int): Int {
      var count = 0
      var idx = beginIndex
      while (idx < endIndex) {
        val ch = seq[idx++]
        if (ch.isHighSurrogate() && idx < endIndex && seq[idx].isLowSurrogate()) {
          // skip the second char of surrogate pairs
          ++idx
        }
        ++count
      }
      return count
    }

    fun codePointCount(ca: CharArray, offset: Int, count: Int): Int {
      val endIndex = offset + count
      var result = 0
      var idx = offset
      while (idx < endIndex) {
        val ch = ca[idx++]
        if (ch.isHighSurrogate() && idx < endIndex && ca[idx].isLowSurrogate()) {
          // skip the second char of surrogate pairs
          ++idx
        }
        ++result
      }
      return result
    }

    fun isValidCodePoint(codePoint: Int): kotlin.Boolean =
      MIN_CODE_POINT <= codePoint && codePoint <= MAX_CODE_POINT

    fun isSurrogatePair(high: Char, low: Char): kotlin.Boolean =
      high.isHighSurrogate() && low.isLowSurrogate()

    fun isLowSurrogate(ch: Char): kotlin.Boolean = ch.isLowSurrogate()

    fun isHighSurrogate(ch: Char): kotlin.Boolean = ch.isHighSurrogate()

    fun isSurrogate(ch: Char): kotlin.Boolean = ch.isSurrogate()

    fun isSupplementaryCodePoint(codePoint: Int) =
      codePoint >= MIN_SUPPLEMENTARY_CODE_POINT && codePoint <= MAX_CODE_POINT

    fun offsetByCodePoints(seq: CharSequence, index: Int, codePointOffset: Int): Int {
      var codePointOffset = codePointOffset
      var index = index
      // move backwards
      while (codePointOffset < 0) {
        --index
        if (seq[index].isLowSurrogate() && seq[index - 1].isHighSurrogate()) {
          --index
        }
        ++codePointOffset
      }
      // move forwards
      while (codePointOffset > 0) {
        if (seq[index].isHighSurrogate() && seq[index + 1].isLowSurrogate()) {
          ++index
        }
        ++index
        --codePointOffset
      }
      return index
    }

    fun offsetByCodePoints(
      ca: CharArray,
      offset: Int,
      count: Int,
      index: Int,
      codePointOffset: Int,
    ): Int {
      var codePointOffset = codePointOffset
      var index = index
      // move backwards
      while (codePointOffset < 0 && index >= 0) {
        --index
        if (ca[offset + index].isLowSurrogate() && ca[offset + index - 1].isHighSurrogate()) {
          --index
        }
        ++codePointOffset
      }
      // move forwards
      while (codePointOffset > 0 && index <= offset + count) {
        if (ca[offset + index].isHighSurrogate() && ca[offset + index + 1].isLowSurrogate()) {
          ++index
        }
        ++index
        --codePointOffset
      }
      if (index < 0 || index > offset + count) {
        throw IndexOutOfBoundsException()
      }
      return index
    }

    fun toCodePoint(high: Char, low: Char): Int =
      (((high - MIN_HIGH_SURROGATE) shl 10) or (low - MIN_LOW_SURROGATE)) +
        MIN_SUPPLEMENTARY_CODE_POINT

    fun toChars(codePoint: Int): CharArray =
      if (codePoint >= MIN_SUPPLEMENTARY_CODE_POINT)
        charArrayOf(highSurrogate(codePoint), lowSurrogate(codePoint))
      else charArrayOf(codePoint.toChar())

    fun toChars(codePoint: Int, dst: CharArray, dstIndex: Int): Int {
      // TODO(b/228304843): Add InternalPreconditions for Kotlin.
      // checkCriticalArgument(Char.isValidCodePoint(codePoint));

      if (codePoint >= MIN_SUPPLEMENTARY_CODE_POINT) {
        dst[dstIndex] = highSurrogate(codePoint)
        dst[dstIndex + 1] = lowSurrogate(codePoint)
        return 2
      }
      dst[dstIndex] = codePoint.toChar()
      return 1
    }

    fun toString(c: Char): kotlin.String = c.toString()

    fun toString(c: Int): kotlin.String = toChars(c).concatToString()

    fun toUpperCase(c: Char) = c.uppercaseChar()

    fun toUpperCase(c: Int): Int = toString(c).uppercase().codePointAt(0)

    fun toLowerCase(c: Char) = c.lowercaseChar()

    fun toLowerCase(c: Int): Int = toString(c).lowercase().codePointAt(0)

    fun toTitleCase(c: Char) = c.titlecaseChar()

    fun lowSurrogate(codePoint: Int) =
      (MIN_LOW_SURROGATE + ((codePoint - MIN_SUPPLEMENTARY_CODE_POINT) and 1023)).toChar()

    fun highSurrogate(codePoint: Int) =
      (MIN_HIGH_SURROGATE + (((codePoint - MIN_SUPPLEMENTARY_CODE_POINT) shr 10) and 1023)).toChar()
  }
}
